File: sdk/js/src/protocol.js
import { ulid } from 'ulid';
import { sign } from './crypto.js';

/**
 * Build and sign a protocol message.
 * @param {string} type - Message type (rfq, bid, accept, cancel, receipt)
 * @param {string} from - Sender identity
 * @param {string} secretKey - Base64-encoded Ed25519 secret key
 * @param {Object} payload - Type-specific fields
 * @param {string|null} [ref] - Parent message ID
 * @param {number} [ttl] - Time to live in seconds
 * @param {string|null} [to] - Target agent (null for broadcast)
 * @returns {Object} Signed protocol message
 */
export function makeMessage(type, from, secretKey, payload, ref = null, ttl = 30, to = null) {
  const body = {
    proto: 'intent/0.1',
    type,
    id: ulid(),
    ref,
    from,
    ...(to && { to }),
    ts: Math.floor(Date.now() / 1000),
    ttl,
    ...payload,
  };
  body.sig = sign(body, secretKey);
  return body;
}

/**
 * Create a signed RFQ message.
 * @param {string} from - Agent identity
 * @param {string} secretKey - Base64 secret key
 * @param {import('./types.js').Intent} intent - Intent payload
 * @param {number} [ttl=30] - TTL in seconds
 * @returns {import('./types.js').RFQMessage}
 */
export function makeRFQ(from, secretKey, intent, ttl = 30) {
  return makeMessage('rfq', from, secretKey, { intent }, null, ttl);
}

/**
 * Create a signed Bid message.
 * @param {string} from - Agent identity
 * @param {string} secretKey - Base64 secret key
 * @param {string} rfqId - ID of the RFQ being bid on
 * @param {import('./types.js').Offer} offer - The offer
 * @param {import('./types.js').Reputation} [reputation] - Agent reputation
 * @param {string} [to] - Target agent
 * @returns {import('./types.js').BidMessage}
 */
export function makeBid(from, secretKey, rfqId, offer, reputation, to = null) {
  return makeMessage('bid', from, secretKey, { offer, reputation }, rfqId, 60, to);
}

/**
 * Create a signed Accept message.
 * @param {string} from - Agent identity
 * @param {string} secretKey - Base64 secret key
 * @param {string} bidId - ID of the accepted bid
 * @param {import('./types.js').Settlement} [settlement] - Settlement terms
 * @returns {Object}
 */
export function makeAccept(from, secretKey, bidId, settlement) {
  return makeMessage('accept', from, secretKey, { accepted_bid: bidId, settlement }, bidId, 10);
}

/**
 * Create a signed Cancel message.
 * @param {string} from - Agent identity
 * @param {string} secretKey - Base64 secret key
 * @param {string} refId - ID of the message being cancelled
 * @param {string} [reason] - Cancellation reason
 * @returns {Object}
 */
export function makeCancel(from, secretKey, refId, reason) {
  return makeMessage('cancel', from, secretKey, { reason, within_terms: true }, refId, 10);
}

/**
 * Create a signed Receipt message.
 * @param {string} from - Agent identity
 * @param {string} secretKey - Base64 secret key
 * @param {string} dealId - Deal ID being confirmed
 * @param {Object} fulfillment - Fulfillment details
 * @returns {Object}
 */
export function makeReceipt(from, secretKey, dealId, fulfillment) {
  return makeMessage('receipt', from, secretKey, { fulfillment }, dealId, 0);
}

File: sdk/python/intentprotocol/protocol.py
"""Intent Protocol message builders."""

from __future__ import annotations

import time
import uuid
from .crypto import sign


def _ulid() -> str:
    """Generate a unique message ID (UUID4-based for simplicity)."""
    return str(uuid.uuid4()).replace("-", "").upper()[:26]


def make_message(
    type_: str,
    from_: str,
    secret_key: bytes,
    payload: dict,
    ref: str | None = None,
    ttl: int = 30,
    to: str | None = None,
) -> dict:
    """Build and sign a protocol message.

    Args:
        type_: Message type (rfq, bid, accept, cancel, receipt)
        from_: Sender identity
        secret_key: Ed25519 secret key (64 bytes)
        payload: Type-specific fields
        ref: Parent message ID
        ttl: Time to live in seconds
        to: Target agent (None for broadcast)

    Returns:
        Signed message dict
    """
    body = {
        "proto": "intent/0.1",
        "type": type_,
        "id": _ulid(),
        "ref": ref,
        "from": from_,
        "ts": int(time.time()),
        "ttl": ttl,
        **payload,
    }
    if to:
        body["to"] = to

    body["sig"] = sign(body, secret_key)
    return body


def make_rfq(from_: str, secret_key: bytes, intent: dict, ttl: int = 30) -> dict:
    """Create a signed RFQ message."""
    return make_message("rfq", from_, secret_key, {"intent": intent}, ttl=ttl)


def make_bid(
    from_: str,
    secret_key: bytes,
    rfq_id: str,
    offer: dict,
    reputation: dict | None = None,
    to: str | None = None,
) -> dict:
    """Create a signed Bid message."""
    payload = {"offer": offer}
    if reputation:
        payload["reputation"] = reputation
    return make_message("bid", from_, secret_key, payload, ref=rfq_id, ttl=60, to=to)


def make_accept(
    from_: str,
    secret_key: bytes,
    bid_id: str,
    settlement: dict | None = None,
) -> dict:
    """Create a signed Accept message."""
    payload = {
        "accepted_bid": bid_id,
        "settlement": settlement or {"method": "direct", "pay_at": "on_site"},
    }
    return make_message("accept", from_, secret_key, payload, ref=bid_id, ttl=10)


def make_cancel(
    from_: str,
    secret_key: bytes,
    ref_id: str,
    reason: str | None = None,
) -> dict:
    """Create a signed Cancel message."""
    return make_message(
        "cancel", from_, secret_key, {"reason": reason, "within_terms": True}, ref=ref_id, ttl=10,
    )


def make_receipt(
    from_: str,
    secret_key: bytes,
    deal_id: str,
    fulfillment: dict | None = None,
) -> dict:
    """Create a signed Receipt message."""
    return make_message(
        "receipt", from_, secret_key, {"fulfillment": fulfillment or {"completed": True}},
        ref=deal_id, ttl=0,
    )
